xls/public/c_api_dslx.cc

```diff
@@ -341,4 +432,125 @@ bool xls_dslx_type_is_signed_bits(const struct xls_dslx_type* type,
   return true;
 }
 
+bool xls_dslx_type_is_enum(const struct xls_dslx_type* type) {
+  const auto* cpp_type = reinterpret_cast<const xls::dslx::Type*>(type);
+  return cpp_type->IsEnum();
+}
+
+bool xls_dslx_type_is_struct(const struct xls_dslx_type* type) {
+  const auto* cpp_type = reinterpret_cast<const xls::dslx::Type*>(type);
+  return cpp_type->IsStruct();
+}
+
+bool xls_dslx_type_is_array(const struct xls_dslx_type* type) {
+  const auto* cpp_type = reinterpret_cast<const xls::dslx::Type*>(type);
+  return cpp_type->IsArray();
+}
+
+struct xls_dslx_type* xls_dslx_type_array_get_element_type(struct xls_dslx_type* type) {
+  const auto* cpp_type = reinterpret_cast<const xls::dslx::Type*>(type);
+  CHECK(cpp_type->IsArray());
```

Comment:
q: should we check and return `nullptr` instead similar to the other function (or use `CHECK` everywhere?)

Reply:
Yeah, there's 3-4 ways to approach APIs like this:
- Defensive -- try to detect everything that can go wrong and return null or an error via an expanded signature -- for functions that are documented as fallible we often return the result as an outparam for this
- Asserting -- document preconditions and try to CHECK them
- Debug-build Asserting -- document preconditions and DCHECK them
- Lax -- document preconditions and assume they've been appropriately followed

I'd say these APIs have been either Asserting or Lax, only trend towards "defensive" is there when the API signatures are expanded to e.g. propagate back absl::Status data.

LMK if that makes sense.

Reply:
Yep, thanks for taking the time to spell that out.